File: 110.R
# Diophantine reciprocals II

options(scipen=100)

Args  <- commandArgs()
limit <- ifelse(is.na(Args[6]), 4e6, as.numeric(Args[6]))

get_primes_up_to <- function(limit) {
    sieve_bound <- (limit - 1) %/% 2 # Last index of sieve
    sieve <- rep(FALSE, sieve_bound)
    cross_limit <- (floor(sqrt(limit)) - 1) %/% 2

    i <- 1
    while ( i <= cross_limit ) {
        if (!sieve[i]) {
            # 2 * $i + 1 is prime, so mark multiples
            j <- 2 * i * (i + 1)
            while ( j <= sieve_bound ) {
                sieve[j] <- TRUE
                j <- j + 2 * i + 1
            }
        }
        i <- i + 1
    }

    primes <- rep(0, sieve_bound)
    for ( i in seq(1, sieve_bound) ) {
        if ( !sieve[i] ) {
            primes[i] <- 2 * i + 1
        }
    }

    return(c(2, primes[primes > 0]))
}

primes <- get_primes_up_to(sqrt(limit))

max_primes <- ceiling(log(2 * limit, 3))
primes <- primes[1:max_primes]

exponent_set <- c(1)
min_n2 <- NA
prev_min_n2 <- NA
while ( is.na(prev_min_n2) || min_n2 != prev_min_n2 ) {
    prev_min_n2 <- min_n2
    exponent_set <- c(exponent_set, exponent_set[length(exponent_set)] + 2)
    idx <- rep(1, max_primes)
    while ( TRUE ) {
        finished <- TRUE
        for ( i in rev(seq.int(max_primes)) ) {
            if ( idx[i] != length(exponent_set) ) {
                finished <- FALSE
                break
            }
        }
        if ( finished ) {
            break
        }
        idx[i:length(idx)] <- rep(idx[i] + 1, max_primes - i + 1)
        product <- prod(exponent_set[idx])
        if ( product > 2 * limit ) {
            exponents <- rev(exponent_set[idx])
            n2 <- 1
            for ( j in seq.int(max_primes) ) {
                n2 <- n2 * primes[j] ^ (exponents[j] - 1)
            }
            if ( !is.na(min_n2) && n2 >= min_n2 ) {
                next
            }
            min_n2 <- n2
        }
    }
}

cat(sqrt(min_n2), fill=TRUE)

File: 110.js
#!/usr/bin/env node

// Diophantine reciprocals II

var program = require('commander');

program
    .version('0.1.0')
    .description('Diophantine reciprocals II')
    .option('-l, --limit <int>', 'The minimum number of distinct solutions',
        Number, 4000000)
    .parse(process.argv);

var primes = get_primes_up_to(Math.sqrt(program.limit));
var max_primes = Math.ceil(Math.log(2 * program.limit) / Math.log(3));
primes = primes.slice(0, max_primes);

var exponent_set = [1];
var min_n2;
var prev_min_n2;
while (typeof prev_min_n2 == 'undefined' || min_n2 != prev_min_n2) {
    prev_min_n2 = min_n2;
    exponent_set.push(exponent_set[exponent_set.length - 1] + 2);
    var idx = new Array(max_primes + 1).join('0').split('').map(parseFloat);
    while (true) {
        var finished = true;
        for (var i = max_primes - 1; i >= 0; i--) {
            if (idx[i] != exponent_set.length - 1) {
                finished = false;
                break;
            }
        }
        if (finished) {
            break;
        }
        var next = idx[i] + 1;
        for (var j = i; j < idx.length; j++) {
            idx[j] = next;
        }
        var product = 1;
        for (var j = 0; j < max_primes; j++) {
            product *= exponent_set[idx[j]];
        }
        if (product > 2 * program.limit) {
            var exponents = [];
            for (var j = 0; j < max_primes; j++) {
                exponents.push(exponent_set[idx[max_primes - j - 1]]);
            }
            var n2 = 1;
            for (var j = 0; j < max_primes; j++) {
                n2 *= Math.pow(primes[j], (exponents[j] - 1));
            }
            if (typeof min_n2 != 'undefined' && n2 >= min_n2) {
                continue;
            }
            min_n2 = n2;
        }
    }
}

console.log(Math.sqrt(min_n2));

function get_primes_up_to(limit) {
    var sieve_bound = Math.floor((limit - 1) / 2); // Last index of sieve
    var sieve = [];
    var cross_limit = Math.floor((Math.floor(Math.sqrt(limit)) - 1) / 2);
    for (var i = 1; i <= cross_limit; i++) {
        if (!sieve[i - 1]) {
            // 2 * i + 1 is prime, so mark multiples
            var j = 2 * i * (i + 1);
            while (j <= sieve_bound) {
                sieve[j - 1] = true;
                j += 2 * i + 1;
            }
        }
    }

    var primes = [2];
    for (var i = 1; i <= sieve_bound; i++) {
        if (!sieve[i - 1]) {
            primes.push(2 * i + 1);
        }
    }

    return primes;
}
